### PR TITLE
Crafting uistate plus refactoring

### PR DESCRIPTION
**Pull request by PatrikStridvall**
Made the crafting dialog remember the tab and the subtab plus refactored the big bloated uistate and and moved the code and data structures closer to where they are used.

Used pimpls and adapters so only required compilation units that needs to known the exact type of each uistate data structure.

### CODE REVIEW

File: src/construction.cpp [hunk 1]
```diff
@@ -1247,6 +1247,18 @@ void check_constructions()
     }
 }
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
```
Comment (Coolthulhu) [inline src/construction.cpp]: Can't you just change the signature to
```
void construction_uistatedata::serialize(JsonOut &) const
```
?

Comment (PatrikStridvall) [reply src/construction.cpp]: I agree, after looking in the repository more carefully as well as doing some testing and running make astyle, it seems that preferred way of this repository is
```
( void ) json; // unused
```

I have refactored my old branch and made a new one (https://github.com/PatrikStridvall/Cataclysm-DDA/tree/crafting-uistate-plus-refactoring2). Not sure  how to update this Pull Request. Will now try to find out.

Comment (Coolthulhu) [reply src/construction.cpp]: > I hope that this doesn't confuse the automatic build process.. or the merge for that matter.

It's fine most of the time. When it's not, you could try the magical phrase "Jenkins, rebuild" which forces the bot to requeue the PR. Or just make a dummy commit.

**Comment by illi-kun** [comment]
Jenkins rebuild

**Comment by PatrikStridvall** [comment]
Because of reported merge problems. I have rebased and forced push over the old branch. I just point this out since the rebasing doesn't update the dates automatically  so it seems like the code is 7 days even though it have been updated.

In the future do you want rebases and forced pushes over the old branch to have updated dates or it is OK to just rebase and fix the merge problems and keep to old date? Or do you want to clearly see that it really has been updated?

**Comment by ZhilkinSerg** [comment]
Jenkins rebuild.

**Comment by Leland** [comment]
Merges are preferred to rebases.

If you rebase after a reviewer has left a note, their review will no longer be tied to a commit and it's very difficult to see what's been changed in your new code

> so it seems like the code is 7 days even though it have been updated

This isn't a problem

**Comment by BevapDin** [comment]
This shows that we don't need a *centralized* ui state class. Each part of the code (that needs to store persistent ui data) can do that on its own, it only needs to expose a `serialize` and `deserialize` function. What is the point of exposing the state of the advanced inventory to *the world*?

File: src/uistate.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    modules.emplace_back( new JsonSerDesAdapter<overmap_uistatedata>( *overmap ) );
+    modules.emplace_back( new JsonSerDesAdapter<player_uistatedata>( *player ) );
+    modules.emplace_back( new JsonSerDesAdapter<wish_uistatedata>( *wish ) );
+}
+
+void uistatedata::serialize( JsonOut &json ) const
+{
+    json.start_object();
+
+    for( auto &module : modules ) {
```
[the remark this answers is not in the document]
Comment (PatrikStridvall) [reply src/uistate.cpp]: This problem already exist in the current code. However most members have a prefix and those that don't probably should have.

In any case I don't want to change the current save format and when we do each module can get an object of each own and the prefix removed.

In short, fixing this is premature.

**Comment by PatrikStridvall** [comment]
@BevapDin Missed your second reply before doing the first. I agree. Sure we could go all way and make uistate.json a directory instead of a file.

This patch splits and keeps a common file save file which is good first step. Changing the saveformat we can to do later and when we do that centralized uistate can be removed.

**Comment by AMurkin** [comment]
Jenkins rebuild.

**Comment by PatrikStridvall** [comment]
@illi-kun Jekins says "To run this automated build, certain files must be exactly as in master:" and then it complains that I have modified src/json.h. The is deliberate and not an error. So Jenkins is "wrong", not my code.

In short, nothing to review.

**Comment by ZhilkinSerg** [comment]
You should not change `json.h` and `json.cpp` files. Move your new classes from these files elsewhere.

See comments below https://github.com/CleverRaven/Cataclysm-DDA/pull/22701#issuecomment-355755035

**Comment by PatrikStridvall** [comment]
@ZhilkinSerg As you wish. Now all checks are green.

**Comment by ZhilkinSerg** [comment]
> As you wish. Now all checks are green.

That is not my wish :)

**Comment by ZhilkinSerg** [comment]
Feel free to reopen.

**Pull request closed without merging**